File: snake-strategy-bono/src/main/java/co/electric/snake/strategy/bonostrategy/directionprocessor/filtereddirections/FilteredDirectionsProcessor.java
package co.electric.snake.strategy.bonostrategy.directionprocessor.filtereddirections;

import co.electric.snake.framework.model.Arena;
import co.electric.snake.framework.model.Coordinate;
import co.electric.snake.framework.model.Snake;
import co.electric.snake.strategy.bonostrategy.SimpleDirectionContainer;
import co.electric.snake.strategy.bonostrategy.directionprocessor.filtereddirections.closeddirections.ClosedDirectionsProcessor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class FilteredDirectionsProcessor {

    private static final Logger LOG = LoggerFactory.getLogger(FilteredDirectionsProcessor.class);

    private final FreeDirectionsProcessor freeDirectionsProcessor;
    private final ClosedDirectionsProcessor closedDirectionsProcessor;

    public FilteredDirectionsProcessor(FreeDirectionsProcessor freeDirectionsProcessor, ClosedDirectionsProcessor closedDirectionsProcessor) {
        this.freeDirectionsProcessor = freeDirectionsProcessor;
        this.closedDirectionsProcessor = closedDirectionsProcessor;
    }

    public SimpleDirectionContainer getDirections(Snake snake, Arena arena) {
        Coordinate actualHeadCoordinate = snake.getHeadCoordinate();

        SimpleDirectionContainer freeDirections = freeDirectionsProcessor.getDirections(arena, actualHeadCoordinate);
        SimpleDirectionContainer closedDirections = closedDirectionsProcessor.getDirections(arena, actualHeadCoordinate);

        SimpleDirectionContainer filteredDirections = freeDirections.getElementsInANewInstance();
        filteredDirections.removeAll(closedDirections);

        LOG.info("Filtered Directions: " + filteredDirections);

        return filteredDirections;
    }

}

File: snake-strategy-bono/src/main/kotlin/co/electric/snake/strategy/bonostrategy/directionprocessor/filtereddirections/FilteredDirectionsProcessor.kt
package co.electric.snake.strategy.bonostrategy.directionprocessor.filtereddirections

import co.electric.snake.framework.model.Arena
import co.electric.snake.framework.model.Snake
import co.electric.snake.strategy.bonostrategy.SimpleDirectionContainer
import co.electric.snake.strategy.bonostrategy.directionprocessor.filtereddirections.closeddirections.ClosedDirectionsProcessor
import org.slf4j.LoggerFactory

class FilteredDirectionsProcessor(private val freeDirectionsProcessor: FreeDirectionsProcessor, private val closedDirectionsProcessor: ClosedDirectionsProcessor) {

    companion object {
        private val LOG = LoggerFactory.getLogger(FilteredDirectionsProcessor::class.java)
    }

    fun getDirections(snake: Snake, arena: Arena): SimpleDirectionContainer {
        val headCoordinate = snake.getHeadCoordinate()
        val freeDirections = freeDirectionsProcessor.getDirections(arena, headCoordinate)
        val closedDirections = closedDirectionsProcessor.getDirections(arena, headCoordinate)
        val filteredDirections = freeDirections.getElementsInANewInstance()
        filteredDirections.removeAll(closedDirections)
        LOG.info("Filtered Directions: $filteredDirections")
        return filteredDirections
    }

}
